flips/20220111-execution-effort.md

```diff
@@ -0,0 +1,555 @@
+
+#### Weight correlation
+
+Before using the data in a linear model, its important to note that the rank of the matrix of all weight data <!-- $M$ --> <img style="transform: translateY(0.1em); background: white;" src="./20220111-execution-effort/eq/SmblEwKco6.svg"> is smaller than the number of weights. This means that some wights are correlated to others. These are asy to spot by checking the correlation matrix where only the entries with 1 or -1 are highlighted.
+
+![Correlation1](./20220111-execution-effort/correlation1.png)
+
+The pairs that are correlated are:
+
+- `ProgramChecked` / `ProgramParsed`
+- `GetCode` / `GetAccountContractCode`
+- `ValueExists` / `CreateAccount`
+
+These pairs of weights were cross checked with the implementation to confirm that they are always called together.
+
+The columns with weights `ProgramChecked`, `GetCode` and `ValueExists` were removed from the weight data matrix <!-- $M$ --> <img style="transform: translateY(0.1em); background: white;" src="./20220111-execution-effort/eq/SmblEwKco6.svg">.
+
+#### Outliers and robust linear model fitting
+
+The following is a linear model fit on the data and plotting the data on a graph of execution time taken (in milliseconds) vs execution effort.
+
+![DemoFit](./20220111-execution-effort/demo-fit.png)
+
+This data has a proportional relation, but also has a lot of noise especially to the right of the graph. Noise to the right of the graph means that sometimes transactions take longer than expected (which could be due to gc pauses, or the machine doing something else, or cache misses, ...).
+
+The outliers cause the goodness of the linear model fit to degrade. A way how to remedy this is described in detail in the [Wolfram documentation](https://reference.wolfram.com/applications/eda/RobustFitting.html) and can be described in short as using the residual (error) of each data point to tweak the weight of that data point, and then re-fitting the data with regard to those weights. This is done until the model converges. In the wolfram documentation this is called robust fitting.
+
+The function that converts the residual of a data point to its weight was chosen to be an asymmetrical function with a different cut-off to the left and to the right. This is because the outliers were mostly to the right of the graph, and the signal was to the left of the graph. The cut-off point was 16 times of the mean of the residuals while the cut-off point to the right was 10 times of the mean residual.
+
+![Residual2Weight](./20220111-execution-effort/residual-to-weight.png)
```

Comment:
Where does this graph come from? Sorry, I don't understand. Who determined that this is exactly the shape we wanted to use?

Reply:
This relates to my reply here https://github.com/onflow/flow/pull/753#discussion_r812110706. I tried to capture a special property of the errors of execution time.